text/0193-fixing-type-unions.md

```diff
@@ -0,0 +1,604 @@
+**JavaScript** runtimes would continue executing).
+
+# Rationale and Alternatives
+
+Improving support for _type unions_ appears to be the best way going forward, as complete removal of type unions has proven to be challenging from an
+API design pespective: it tends to make the experience significantly worse for languages with antive support for _type unions_ (as boilerplate-heavy
+patterns must be leveraged as a replacement for _type unions_), and would reduce the flexibility in language-specific code generation for _union-like_
+types (as those would be bound to the **TypeScript** API). Full support of _type unions_ as a type kind in the _jsii type model_ achieves better
+outcomes:
+
+- it ensures consistency around _union-like_ types
+- it does not degrade the developer experience in languages with _native type_ union support
+- it requires significantly less effort from library authors
+
+# Adoption Strategy
+
+Leveraging aliased _type unions_ in all existing CDK APIs would be a breaking change for all bindings to languages without native _type unions_. As
+such, the initial rollout will start as early as possible in AWS CDK v1, however existing APIs will not be updated to leverage aliased _type unions_,
+preserving all existing APIs.
+
+If the solution is ready on time, it may be possible to update all existing APIs to aliased _type unions_ as part of the [AWS CDK v2] release,
+although this is not a hard requirement: those APIs that are currently (in AWS CDK v1) unusable in **C#** and **Java** can be intentionally updated
+outside of the context of a new major version, since existing code is already broken.
+
+[aws cdk v2]: https://github.com/aws/aws-cdk-rfcs/issues/79
+
+# Unresolved questions
+
+- Are there type union patterns that will be problematic when generating union-like types?
+- Are there better API patterns we could leverage in various languages?
+
+# Future Possibilities
+
+It will be possible in the future to make the `jsii` compiler **require** aliased _type unions_, ensuring the best possible experience for developers
+in languages without native _type unions_. This can either be introduced in the current version of `jsii` though a new option (e.g:
+`--strict=type-union`); alternatively, a new major version could be released that does **not** support inline _type unions_ at all.
+
+As `jsii` gained the ability to track _type aliases_, more use for type aliasing can be considered valid. In particular, certain aliasing can be
+leveraged to disambiguate class names in certain contexts. For example, when renaming a class as part of a refactoring, a `@deprecated` type alias can
+be provided as a way to maintain compatibility with the old name at very little cost.
+
+# Implementation Plan
+
+- [ ] Implement support for aliased _type unions_ in `jsii`
+- [ ] Implement _union-like_ type emission in `jsii-pacmak`
+  - [ ] in **C#**
+  - [ ] in **Java**
+  - [ ] in **Go**
```

Comment:
Still need to do something for Python. As specced, an object literal will contain the FQN for the union type, which Python will need to handle. Although if we change the spec I *think* we won't need to do more for Python than handle the union type in pacmak and ignore it :).

Reply:
> Still need to do something for Python. As specced, an object literal will contain the FQN for the union type, which Python will need to handle

This is naturally required as new tests are added to validate those APIs can be used and behave in the expected way.

> Although if we change the spec I think we won't need to do more for Python than handle the union type in pacmak and ignore it :)

This was my intention (aka de-sugar the type aliases in Python).